bgp: allow duplicated bgppeer per neighbor ip

clwluvw opened the pull request:
**Is this a BUG FIX or a FEATURE ?**:

/kind feature

**What this PR does / why we need it**:

Ensure that duplicate BGP peers in FRR mode have nodeSelectors to distinguish them.
Also, add a runtime check in the BGP controller to detect duplicate peers by address or interface, and return a clear error if found.

Fixes: https://github.com/metallb/metallb/issues/2367

**Release note**:
```release-note
Allow duplicated BGPPeer per neighbor IP
```

clwluvw committed:
bgp: allow duplicated bgppeer per neighbor ip

Ensure that duplicate BGP peers in FRR mode have nodeSelectors to
distinguish them.
Also add a runtime check in the BGP controller to detect duplicate
peers by address or interface, and return a clear error if found.

Signed-off-by: Seena Fallah <seenafallah@gmail.com>

clwluvw commented:
ping @fedepaol @oribon :)

clwluvw commented:
ping @fedepaol @oribon :)

fedepaol commented on internal/config/config.go:
```diff
@@ -252,9 +252,6 @@ func For(resources ClusterResources, validate Validate) (*Config, error) {
 	}
 
 	cfg.BGPExtras = bgpExtrasFor(resources)
-	if err != nil {
```
makes sense, but not related to this change, please add it in its own commit

fedepaol commented on internal/config/validation.go:
```diff
@@ -146,19 +146,22 @@ func DontValidate(c ClusterResources) error {
 			peerID := peerIdentifier(p.Spec)
-			if _, ok := peerAddr[peerID]; ok {
-				return fmt.Errorf("peer %s already exists, FRR mode doesn't support duplicate BGPPeers", p.Spec.Address)
+			if hadNodeSelectors, ok := peerAddrHasNodeSelectors[peerID]; ok {
```
checking if the other peer has a selector is not enough, we should ensure that there are no overlapping nodes, and not sure we can do that at this stage

fedepaol replied:
sorry if I reply so late. I think we are rejecting multiple peers with same asn number / ip address, so it should be fine to ensure they are not configured on overlapping nodes

clwluvw replied:
Thanks @fedepaol. I've simplified validateDuplicatePeers to just reject any duplicate peers whose node selectors might overlap, and removed the arePeersCompatible check entirely.

fedepaol commented:
sorry for the delay, left some comments.

Also, the hardest part here is the validation. We need to check for the validity of each node, but we can't prevent a valid configuration become invalid because somebody changes a label to a node (and there's not much we can do about that, because we certainly don't want to add a webhook over the node resource)

fedepaol commented:
also, ignore copilot's comments

clwluvw committed:
config: remove always false if branch

Signed-off-by: Seena Fallah <seenafallah@gmail.com>

clwluvw committed:
config: validate duplicate BGP peers with node selector compatibility

Duplicate peers are permitted when they have node selectors that
either clearly don't overlap or when they have compatible
configurations that would result in identical BGP sessions.
Two peers are considered compatible when all their session
parameters match except for the node selectors themselves.

If it cannot definitively prove that node selectors are disjoint,
it requires the peers to be compatible.

Signed-off-by: Seena Fallah <seenafallah@gmail.com>

clwluvw committed:
speaker: consolidate peer filtering and deduplication logic

This refactoring extracts peer filtering and deduplication into
a single function that is called consistently from both SetConfig
and SetNode. Previously, this logic was scattered across multiple
places with syncPeers doing filtering based on node selectors while
also attempting duplicate detection.

Signed-off-by: Seena Fallah <seenafallah@gmail.com>

clwluvw commented:
Unstable me please

clwluvw commented:
unstale me

clwluvw commented:
unstale me

clwluvw committed:
validation: reject duplicate peers with overlapping node selectors

Remove the arePeersCompatible check from duplicate peer validation.
Since peers with different myASN in the same VRF are already rejected,
duplicate peers (same address/interface + VRF) cannot have meaningfully
different but valid configurations. Simplify to always reject duplicates
whose node selectors might overlap, regardless of config similarity.

Signed-off-by: Seena Fallah <seenafallah@gmail.com>

fedepaol commented:
Superseeded by https://github.com/metallb/metallb/pull/2980
